File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/prefer-to-have-been-called-times.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('prefer-to-have-been-called-times', {} as never, {
  valid: [
    { code: 'expect.assertions(1)' },
    { code: 'expect(fn).toHaveBeenCalledTimes' },
    { code: 'expect(fn.mock.calls).toHaveLength' },
    { code: 'expect(fn.mock.values).toHaveLength(0)' },
    { code: 'expect(fn.values.calls).toHaveLength(0)' },
    { code: 'expect(fn).toHaveBeenCalledTimes(0)' },
    { code: 'expect(fn).resolves.toHaveBeenCalledTimes(10)' },
    { code: 'expect(fn).not.toHaveBeenCalledTimes(10)' },
    { code: 'expect(fn).toHaveBeenCalledTimes(1)' },
    { code: 'expect(fn).toBeCalledTimes(0);' },
    { code: 'expect(fn).toHaveBeenCalledTimes(0);' },
    { code: 'expect(fn);' },
    { code: 'expect(method.mock.calls[0][0]).toStrictEqual(value);' },
    { code: 'expect(fn.mock.length).toEqual(1);' },
    { code: 'expect(fn.mock.calls).toEqual([]);' },
    { code: 'expect(fn.mock.calls).toContain(1, 2, 3);' },
    { code: 'expect((fn.mock.calls)).toEqual([]);' },
  ],
  invalid: [
    {
      code: 'expect(method.mock.calls).toHaveLength(1);',
      output: 'expect(method).toHaveBeenCalledTimes(1);',
      errors: [
        {
          messageId: 'preferMatcher',
          column: 27,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(method.mock.calls).resolves.toHaveLength(x);',
      output: 'expect(method).resolves.toHaveBeenCalledTimes(x);',
      errors: [
        {
          messageId: 'preferMatcher',
          column: 36,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(method["mock"].calls).toHaveLength(0);',
      output: 'expect(method).toHaveBeenCalledTimes(0);',
      errors: [
        {
          messageId: 'preferMatcher',
          column: 30,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(my.method.mock.calls).not.toHaveLength(0);',
      output: 'expect(my.method).not.toHaveBeenCalledTimes(0);',
      errors: [
        {
          messageId: 'preferMatcher',
          column: 34,
          line: 1,
        },
      ],
    },
    {
      code: 'expect((method.mock.calls)).toHaveLength(1);',
      output: 'expect(method).toHaveBeenCalledTimes(1);',
      errors: [
        {
          messageId: 'preferMatcher',
          column: 29,
          line: 1,
        },
      ],
    },
  ],
});

File: internal/plugins/jest/rules/prefer_to_have_been_called_times/prefer_to_have_been_called_times_test.go
package prefer_to_have_been_called_times_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/prefer_to_have_been_called_times"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestPreferToHaveBeenCalledTimesRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&prefer_to_have_been_called_times.PreferToHaveBeenCalledTimesRule,
		[]rule_tester.ValidTestCase{
			{Code: `expect.assertions(1)`},
			{Code: `expect(fn).toHaveBeenCalledTimes`},
			{Code: `expect(fn.mock.calls).toHaveLength`},
			{Code: `expect(fn.mock.values).toHaveLength(0)`},
			{Code: `expect(fn.values.calls).toHaveLength(0)`},
			{Code: `expect(fn).toHaveBeenCalledTimes(0)`},
			{Code: `expect(fn).resolves.toHaveBeenCalledTimes(10)`},
			{Code: `expect(fn).not.toHaveBeenCalledTimes(10)`},
			{Code: `expect(fn).toHaveBeenCalledTimes(1)`},
			{Code: `expect(fn).toBeCalledTimes(0);`},
			{Code: `expect(fn).toHaveBeenCalledTimes(0);`},
			{Code: `expect(fn);`},
			{Code: `expect(method.mock.calls[0][0]).toStrictEqual(value);`},
			{Code: `expect(fn.mock.length).toEqual(1);`},
			{Code: `expect(fn.mock.calls).toEqual([]);`},
			{Code: `expect(fn.mock.calls).toContain(1, 2, 3);`},
			{Code: `expect((fn.mock.calls)).toEqual([]);`},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `expect(method.mock.calls).toHaveLength(1);`,
				Output: []string{
					`expect(method).toHaveBeenCalledTimes(1);`,
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 27},
				},
			},
			{
				Code: `expect(method.mock.calls).resolves.toHaveLength(x);`,
				Output: []string{
					`expect(method).resolves.toHaveBeenCalledTimes(x);`,
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 36},
				},
			},
			{
				Code: `expect(method["mock"].calls).toHaveLength(0);`,
				Output: []string{
					`expect(method).toHaveBeenCalledTimes(0);`,
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 30},
				},
			},
			{
				Code: `expect(my.method.mock.calls).not.toHaveLength(0);`,
				Output: []string{
					`expect(my.method).not.toHaveBeenCalledTimes(0);`,
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 34},
				},
			},
			{
				Code: `expect(method.mock.calls).toHaveLength(1, 2);`,
				Output: []string{
					`expect(method).toHaveBeenCalledTimes(1, 2);`,
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 27},
				},
			},
			{
				Code: `expect((method.mock.calls)).toHaveLength(1);`,
				Output: []string{
					`expect(method).toHaveBeenCalledTimes(1);`,
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 29},
				},
			},
		},
	)
}
